hyperspace/cosmos/src/provider.rs

```diff
@@ -0,0 +1,415 @@
+		todo!()
+	}
+
+	async fn query_client_update_time_and_height(
+		&self,
+		client_id: ClientId,
+		client_height: Height,
+	) -> Result<(Height, Timestamp), Self::Error> {
+		todo!()
+	}
+
+	async fn query_host_consensus_state_proof(
+		&self,
+		height: Height,
+	) -> Result<Option<Vec<u8>>, Self::Error> {
+		todo!()
+	}
+
+	async fn query_ibc_balance(&self) -> Result<Vec<PrefixedCoin>, Self::Error> {
+		todo!()
+	}
+
+	fn connection_prefix(&self) -> CommitmentPrefix {
+		todo!()
+	}
+
+	fn client_id(&self) -> ClientId {
+		self.client_id()
+	}
+
+	fn client_type(&self) -> ClientType {
+		todo!()
+	}
+
+	fn connection_id(&self) -> ConnectionId {
+		todo!()
+	}
+
+	async fn query_timestamp_at(&self, block_number: u64) -> Result<u64, Self::Error> {
+		todo!()
+	}
+
+	async fn query_clients(&self) -> Result<Vec<ClientId>, Self::Error> {
+		let request = tonic::Request::new(QueryClientStatesRequest { pagination: None });
+		let grpc_client = ibc_proto::ibc::core::client::v1::query_client::QueryClient::connect(
+			self.grpc_url.clone().to_string(),
+		)
+		.await
+		.map_err(|e| Error::from(format!("failed to connect to grpc client: {:?}", e)))?;
+		let response = grpc_client
+			.clone()
+			.client_states(request)
+			.await
+			.map_err(|e| {
+				Error::from(format!("Failed to query client states from grpc client: {:?}", e))
+			})?
+			.into_inner();
+
+		// Deserialize into domain type
+		let clients: Vec<ClientId> = response
+			.client_states
+			.into_iter()
+			.filter_map(|cs| {
+				let id = ClientId::from_str(&cs.client_id).ok()?;
+				Some(id)
+			})
+			.collect();
+		Ok(clients)
+	}
+
+	async fn query_channels(&self) -> Result<Vec<(ChannelId, PortId)>, Self::Error> {
+		todo!()
+	}
+
+	async fn query_connection_using_client(
+		&self,
+		height: u32,
+		client_id: String,
+	) -> Result<Vec<IdentifiedConnection>, Self::Error> {
+		todo!()
+	}
+
+	fn is_update_required(
+		&self,
+		latest_height: u64,
+		latest_client_height_on_counterparty: u64,
+	) -> bool {
+		todo!()
+	}
+
+	async fn initialize_client_state(
+		&self,
+	) -> Result<(AnyClientState, AnyConsensusState), Self::Error> {
+		let latest_height_timestamp = self.latest_height_and_timestamp().await.unwrap();
+		let client_state = TmClientState::<HostFunctionsManager>::new(
+			self.chain_id.clone(),
+			TrustThreshold::default(),
+			Duration::new(64000, 0),    // Set to a default value
+			Duration::new(128000, 0),   // Set to a default value
+			Duration::new(5, 0),		// Set to a default value
+			latest_height_timestamp.0,
+			ProofSpecs::default(),
+			vec!["upgrade".to_string(), "upgradedIBCState".to_string()],
+		)
+		.map_err(|e| Error::from(format!("Invalid client state {}", e)))?;
+
+		let target_height = TmHeight::try_from(latest_height_timestamp.0.revision_height)
+			.map_err(|e| Error::from(e.to_string()))?;
+		let client = self
+			.construct_tendermint_client_state::<HostFunctionsManager>(&client_state)
+			.await?;
+		let trusted_block =
```

Comment:
Can't we just derive the consensus state directly from this trusted block?

Reply:
Right. That extra verification step shouldn't be needed here. I've refactored it to also comply more with what we have in the `hyperspace/parachain` package. But kept that TmLightClient builder as a separate method `construct_tendemint_light_client()`... just in case needed...